router.py

```diff
@@ -49,41 +48,52 @@ def register_handler(self, from_did, msg_type, handler):
     async def route_message(self, msg):
         msg_type = msg["type"]
         from_did = msg["from"]
-
-        # Create or get context for the sending DID
-        if from_did not in self.contexts:
-            self.contexts[from_did] = {}
-        context = self.contexts[from_did]
+        thid = msg.get("thid")
+
+        # Get or create contact context
+        contact_context = self.context.get_contact_context(from_did)
+        if contact_context is None:
+            contact_context = {}
+            self.context.set_contact_context(from_did, contact_context)
+
+        # Get or create thread context if thid is present
+        thread_context = None
+        if thid:
+            thread_context = self.context.get_thread_context(thid)
+            if thread_context is None:
+                thread_context = {}
+                self.context.set_thread_context(thid, thread_context)
 
         # Check for registered handler
         handler_name = self.named_function_handler.get_handler(from_did, msg_type)
         if handler_name:
             self.named_function_handler.remove_handler(from_did, msg_type)
             handler = self.handler_map.get(handler_name)
             if handler:
-                await self.scheduler.spawn(handler(msg, context))
+                await self.scheduler.spawn(handler(msg, contact_context, thread_context))
```

Comment:
only one context object - choosing contact or thread as a destination is handled by the get and set methods as described in my other comment.